File: C++/48. Rotate Image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        if (matrix.empty())
            return;
        int tmp;
        int m = matrix.size();        
        for(int i = 0; i < m / 2; i++){
            for (int j = i; j < m - i - 1; j++){                
                tmp = matrix[i][j];
                matrix[i][j] = matrix[m - 1 - j][i];
                matrix[m - 1 -j][i] = matrix[m - 1 - i][m - 1 - j];
                matrix[m - 1 - i][m - 1 - j] = matrix[j][m -1 -i];   
                matrix[j][m - 1 - i] = tmp;                
            }            
        }
    }
};

File: Python/48. Rotate Image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        def rotate_layer(r1, c1, r2, c2, matrix):            
            for i in range(0, c2 - c1):
                tmp = matrix[r1][c1 + i]
                matrix[r1][c1 + i] = matrix[r2 - i][c1]      
                matrix[r2 - i][c1] = matrix[r2][c2 - i]
                matrix[r2][c2 - i] = matrix[r1 + i][c2]
                matrix[r1 + i][c2] = tmp                
        if not matrix:
            return        
        r1, r2 = 0, len(matrix) - 1
        c1, c2 = 0, len(matrix[0]) - 1
        while r1 <= r2 and c1 <= c2:
            rotate_layer(r1, c1, r2, c2, matrix)            
            r1 += 1
            r2 -= 1
            c1 += 1
            c2 -= 1        
                
                
                